edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java

```diff
@@ -0,0 +1,256 @@
+package netconsole3;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.util.HashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class RioConnector {
+  private Pattern dsPattern = Pattern.compile("\"robotIP\"[^:]*:[^0-9]*([0-9]+)");
+
+  private final Lock lock = new ReentrantLock();
+  private final Condition cvDone = lock.newCondition();
+  private boolean done = false;
+  private final HashMap<String, Thread> attempts = new HashMap<>();
+  private Socket socket = null;
+
+  private static final int CONNECTION_TIMEOUT_MS = 2000;
+  private static final int TOTAL_TIMEOUT_SEC = 5;
+
+  private ILogger logger;
+
+  public RioConnector(ILogger logger) {
+    this.logger = logger;
+  }
+
+  public Socket connect(int team) throws InterruptedException {
+    if (team <= 0) {
+      return null;
+    }
+    lock.lock();
+    try {
+      done = false;
+      socket = null;
+    } finally {
+      lock.unlock();
+    }
+
+    // start connection attempts to various address possibilities
+    startConnect(new byte[] {10, (byte)(team / 100), (byte)(team % 100), 2});
+    startConnect(new byte[] {(byte)127, 0, 0, 1});
+    startConnect(new byte[] {(byte)172, 22, 11, 2});
+    startConnect("roboRIO-" + team + "-FRC.local");
+    startConnect("roboRIO-" + team + "-FRC.lan");
+    startConnect("roboRIO-" + team + "-FRC.frc-field.local");
+    startDsConnect();
+
+    // wait for a connection attempt to be successful, or timeout
+    lock.lock();
+    try {
+      while (!done && !attempts.isEmpty()) {
+        if (!cvDone.await(TOTAL_TIMEOUT_SEC, TimeUnit.SECONDS)) {
+          //logger.log("Connection timed out.");
+          return null;
+        }
+      }
+
+      if (socket == null) {
+        //logger.log("Not connected to robot");
+        return null;
+      }
+
+      logger.log("Connected to robot at " + socket.getInetAddress().getHostAddress());
+      return socket;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void startDsConnect() {
+    startAttempt("DS", new Thread(() -> {
```

Comment:
You need to make this a daemon thread or the process won't die when the main thread exits.

Reply:
It is. See startAttempt().